src/components/KCodeBlock/KCodeBlock.vue

```diff
@@ -0,0 +1,829 @@
+        v-else
+        class="k-highlighted-code-block"
+        data-testid="k-highlighted-code-block"
+      ><span
+        v-if="props.showLineNumbers"
+        class="k-line-number-rows"
+      >
+        <a
+          v-for="line in totalLines"
+          :id="`${linePrefix}-L${line}`"
+          :key="line"
+          class="k-line"
+          :class="{
+            'k-line--is-match': matchingLineNumbers.includes(line),
+            'k-line--is-highlighted-match': currentLineIndex !== null && line === matchingLineNumbers[currentLineIndex],
+          }"
+          :href="`#${linePrefix}-L${line}`"
+        >{{ line }}</a>
+      </span><code v-html="props.code" /></pre>
+      <!-- eslint-enable vue/no-v-html -->
+
+      <KClipboardProvider
+        v-if="props.showCopyButton"
+        v-slot="{ copyToClipboard }"
+      >
+        <KPop placement="bottom">
+          <KButton
+            class="k-code-block-copy-button"
+            type="button"
+            appearance="outline"
+            :is-rounded="false"
+            size="small"
+            title="Copy"
+            data-testid="k-code-block-copy-button"
+            @click="copyToClipboard(props.code)"
+          >
+            <KIcon
+              icon="copy"
+              size="24"
+              title="Copy"
+              color="currentColor"
+            />
+
+            <span class="k-visually-hidden">Copy</span>
+          </KButton>
+
+          <template #content>
+            Code copied to clipboard
+          </template>
+        </KPop>
+      </KClipboardProvider>
+    </div>
+  </div>
+</template>
+
+<script lang="ts" setup>
+import { computed, nextTick, onBeforeUnmount, onMounted, ref, watch } from 'vue'
+
+import KButton from '@/components/KButton/KButton.vue'
+import KClipboardProvider from '@/components/KClipboardProvider/index'
+import KIcon from '@/components/KIcon/KIcon.vue'
+import KInput from '@/components/KInput/KInput.vue'
+import KLabel from '@/components/KLabel/KLabel.vue'
+import KPop from '@/components/KPop/KPop.vue'
+import { debounce } from '@/utils/debounce'
+
+export type CodeBlockEventData = {
+  preElement: HTMLElement
+  codeElement: HTMLElement
+  code: string
+  language: string
+  query: string
+  matchingLineNumbers: number[]
+}
+
+// Debounces the search handler which ensures that we don’t trigger several searches while the user is still typing.
+const debouncedHandleSearchInputValue = debounce(handleSearchInputValue, 150)
+
+const props = defineProps({
+  /**
+   * ID value used for form elements like the search input and its label.
+   */
+  id: {
+    type: String,
+    required: false,
+    default: 'code-block',
+  },
+
+  /**
+   * The code that will be rendered as a text node inside of a `<pre><code></code></pre>` fragment.
+   */
+  code: {
+    type: String,
+    required: true,
+  },
+
+  /**
+   * The syntax language of `props.code` (e.g. `'json'`).
+   */
+  language: {
+    type: String,
+    required: true,
+  },
```

Comment:
question: should we do any sort of validation on the supported languages here, if any?

Reply:
In the version I implemented in Kuma GUI, I used a more specific type based on the languages I’ve loaded via PrismJS. This isn’t necessary here since this implementation intentionally excludes the responsibility for syntax highlighting. Otherwise, I would’ve needed to introduce PrismJS as a dependency of Kongponents.

In other words: we can’t know which language identifiers a project might end up using and have loaded (projects will certainly have differents needs with regards to which languages they need to support). So for now, `props.language` isn’t actually used _inside_ the component for anything other than emitting it again. It could possibly be used to add the language to the UI.